theano/sandbox/jaxify.py

```diff
@@ -462,63 +468,74 @@ def jax_args_to_inner_scan(op, carry, x):
+                # inner_in_mit_mot,
+                inner_in_mit_sot_flatten,
                 inner_in_sit_sot,
+                inner_in_shared,
                 inner_in_non_seqs,
             ]
 
-            raise NotImplementedError()
             return inner_scan_inputs
 
         def inner_scan_outs_to_jax_outs(
             op,
             old_carry,
             inner_scan_outs,
         ):
-            # `inner_scan_outs` is a list with the following
-            # composite form:
-            # outer_out_mit_mot
-            # + outer_out_mit_sot
-            # + outer_out_sit_sot
-            # + outer_out_nit_sot
-            # + outer_out_shared
-            # + cond
             (
-                outer_out_mit_mot,
                 outer_out_mit_sot,
-                outer_out_sit_sot,
-                outer_out_nit_sot,
-                outer_out_shared,
-                cond,
-            ) = inner_scan_outs
-            outer_out_non_seqs = old_carry[:-n_non_seqs]
+                outer_in_sit_sot,
+                inner_in_shared,
+                outer_in_non_seqs,
+            ) = old_carry
+            if not outer_out_mit_sot:
+                outer_out_mit_sot_next = []
+            else:
+                outer_out_mit_sot_next = jax.tree_multimap(
+                    lambda mit_sot, new_val: jnp.concatenate(
+                        [mit_sot[1:], new_val[None, ...]], axis=0
+                    ),
+                    outer_out_mit_sot,
+                    inner_scan_outs,
+                )
 
             # This should contain all inner-output taps, non_seqs, and shared
             # terms
-            carry = [
-                outer_out_mit_mot,
-                outer_out_mit_sot,
-                outer_out_sit_sot,
-                outer_out_shared,
-                outer_out_non_seqs,
-            ]
-            # This should contain all inner-outputs that produce
-            # outer-outputs
-            y = []
+            if not outer_in_sit_sot:
+                outer_in_sit_sot_next = []
+            else:
+                outer_in_sit_sot_next = inner_scan_outs
+            new_carry = (
+                outer_out_mit_sot_next,
+                outer_in_sit_sot_next,
+                inner_in_shared,
+                outer_in_non_seqs,
+            )
 
-            raise NotImplementedError()
-            return (carry, y)
+            return new_carry
 
         def jax_inner_func(carry, x):
             inner_args = jax_args_to_inner_scan(op, carry, x)
-            inner_scan_outs = jax_tt_inner_func(*inner_args)
-            new_carry, y = inner_scan_outs_to_jax_outs(op, inner_scan_outs)
-            return new_carry, y
-
-        return jax.lax.scan(jax_inner_func, init_carry, seqs, length=n_steps)
+            inner_args = jax.tree_leaves(inner_args)
+            inner_scan_outs = jax.tree_map(
+                lambda fn: fn(*inner_args), jax_tt_inner_func
+            )
```

Comment:
I have problem initially because sometimes functions in `jax_tt_inner_func` is nested in a list, but I think that's fixed by https://github.com/pymc-devs/Theano-PyMC/pull/166

Reply:
As a side note, how do you feel about using more `jax.tree_util` inside of `jaxify.py`. On one hand I like it as it handles nested args/kwargs gracefully, on the other hand in the rest of theano code base we assume things to be a list in many places so it is good to be consistent.

Reply:
I don't know too much about `jax.tree_util` so you'll have to be more specific, but it looks like the module provides a mix of features that overlap with [`etuples`](https://github.com/pythological/etuples) (especially the `rator`, `rands` dispatching and keyword handling) and quite possibly [`cons`](https://github.com/pythological/python-cons).  

I'm a big fan of S-expression emulation in Python, so I can't fundamentally object to this approach; however, it looks like `pytree` source isn't in Python?  If that's true, it doesn't work in its favor, especially since this kind of simple graph work can be done very well (and easily) in standard Python.
 
Otherwise, there aren't many cases in which we handle nested/tree structures in Theano; all graph-related things are represented by the `theano.gof.graph` objects.  As we're seeing here, there are some lists-of-lists, but they don't actually represent graphs.

Now, some of the "JAXification" we're doing involves nested lists, but that's largely an artifact of how a JAXified `FunctionGraph` is represented.  In general, a `FunctionGraph` has a list of outputs, so the JAXified `FunctionGraph` should be a list of JAXable functions for each output.  In general, we shouldn't have to deal with non-Theano representations of graphs.

We could definitely take other approaches to the JAXification process, so, please, ask any and all question that might help clarify how it currently works and&mdash;hopefully&mdash;lead to improvements.  Even if those questions simply result in better documentation, that would be well worth it.

For instance, one of things about our current approach&mdash;and much of what's done in Theano&mdash;that really bothers me is how recursion is handled.  It's done naively in Python and, as such, is bound by the Python stack.  Python has no TCO, so these kinds of problems need to be fixed manually.  At some point, I would like to generalize and use [the approach I took in `unification` and `etuples`](https://github.com/pythological/unification/pull/16): i.e. [a simple generator-based trampoline](https://github.com/pythological/unification/pull/16/files#diff-5675d2e53a975726b4f985153afb1af9fc4d659c3d41868bd4c3604e8fdcfd9dR17).  Last I measured, this approach was about as performant as the naive stack approach for shallow recursions and significantly better for deep recursions (see the tests in the PR linked earlier).

Reply:
I agree with all your points. Indeed nested list should not occur in the current set up, but moving forward it might be useful (thanks for filing the comment). I replace the rest of the `jax.tree_*` for now.

Reply:
Just so it's clear: I don't think there's anything wrong with using those functions.  I was just trying to understand what they did/would do.